### PR TITLE
prevent pad/unpad failures on large/small messages

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: src/Simplex/Messaging/Crypto.hs [hunk 1]
```diff
@@ -691,6 +691,8 @@ data CryptoError
   | -- | message is larger that allowed padded length minus 2 (to prepend message length)
     -- (or required un-padded length is larger than the message length)
     CryptoLargeMsgError
+  | -- | padded message is shorted than 2 bytes
+    CrypteInvalidMsgError
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Crypto.hs:695]: typod
```suggestion
  | -- | padded message is shorter than 2 bytes
    CryptoInvalidMsgError
```

File: src/Simplex/Messaging/Crypto.hs [hunk 2]
```diff
@@ -802,18 +804,21 @@ decryptAEAD aesKey ivBytes ad msg (AuthTag authTag) = do
...
   where
     len = B.length msg
     padLen = paddedLen - len - 2
 
 unPad :: ByteString -> Either CryptoError ByteString
 unPad padded
-  | B.length rest >= len = Right $ B.take len rest
-  | otherwise = Left CryptoLargeMsgError
+  | B.length lenWrd == 2 && B.length rest >= len = Right $ B.take len rest
+  | otherwise = Left CrypteInvalidMsgError
```
Comment (jr-simplex) [inline src/Simplex/Messaging/Crypto.hs:821]:
```suggestion
  | otherwise = Left CryptoInvalidMsgError
```

File: tests/CoreTests/CryptoTests.hs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    it "pad should fail on large string" $ do
+      C.pad "abc" 5 `shouldBe` Right "\000\003abc"
+      C.pad "abc" 4 `shouldBe` Left C.CryptoLargeMsgError
+    it "unpad should fail on invalid string" $ do
+      C.unPad "\000\000" `shouldBe` Right ""
+      C.unPad "\000" `shouldBe` Left C.CrypteInvalidMsgError
+      C.unPad "" `shouldBe` Left C.CrypteInvalidMsgError
+    it "unpad should fail on shorter string" $ do
+      C.unPad "\000\003abc" `shouldBe` Right "abc"
+      C.unPad "\000\003ab" `shouldBe` Left C.CrypteInvalidMsgError
```
Comment (jr-simplex) [inline tests/CoreTests/CryptoTests.hs:34]:
```suggestion
      C.unPad "\000" `shouldBe` Left C.CryptoInvalidMsgError
      C.unPad "" `shouldBe` Left C.CryptoInvalidMsgError
    it "unpad should fail on shorter string" $ do
      C.unPad "\000\003abc" `shouldBe` Right "abc"
      C.unPad "\000\003ab" `shouldBe` Left C.CryptoInvalidMsgError
```

**Review by jr-simplex — APPROVED** [review]

File: src/Simplex/Messaging/Crypto.hs [hunk 4]
```diff
@@ -802,18 +804,21 @@ decryptAEAD aesKey ivBytes ad msg (AuthTag authTag) = do
   aead <- initAEAD @AES256 aesKey ivBytes
   liftEither . unPad =<< maybeError AESDecryptError (AES.aeadSimpleDecrypt aead ad msg authTag)
 
+maxMsgSize :: Int
+maxMsgSize = 2 ^ (16 :: Int) - 3
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Crypto.hs:808]:
```suggestion
maxMsgLen :: Int
maxMsgLen = 2 ^ (16 :: Int) - 3
```

File: src/Simplex/Messaging/Crypto.hs [hunk 5]
```diff
@@ -802,18 +804,21 @@ decryptAEAD aesKey ivBytes ad msg (AuthTag authTag) = do
...
+
 pad :: ByteString -> Int -> Either CryptoError ByteString
 pad msg paddedLen
-  | padLen >= 0 = Right $ encodeWord16 (fromIntegral len) <> msg <> B.replicate padLen '#'
+  | len <= maxMsgSize && padLen >= 0 = Right $ encodeWord16 (fromIntegral len) <> msg <> B.replicate padLen '#'
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Crypto.hs:812]:
```suggestion
  | len <= maxMsgLen && padLen >= 0 = Right $ encodeWord16 (fromIntegral len) <> msg <> B.replicate padLen '#'
```

File: tests/CoreTests/CryptoTests.hs [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+cryptoTests = modifyMaxSuccess (const 10000) $ do
+  describe "padding / unpadding" $ do
+    it "should pad / unpad string" . property $ \(s, paddedLen) ->
+      let b = encodeUtf8 $ T.pack s
+          padded = C.pad b paddedLen
+       in if paddedLen >= B.length b + 2
+            then if B.length b < 2 ^ (16 :: Int) - 3
+              then (fmap (T.unpack . decodeUtf8) . C.unPad =<< padded) == Right s
+              else padded == Left C.CryptoInvalidMsgError
+            else True
```
Comment (epoberezkin) [inline tests/CoreTests/CryptoTests.hs:24]:
```suggestion
            else padded == Left C.CryptoLargeMsgError
```

**Comment by remonshikder99-jpg** [comment]
@epoberezkin
@spaced4ndy
[typos](https://github.com/simplex-chat/simplexmq/pull/547/commits/9b16648cfe3c694302ad706cd9d154fec845f3ec) 
[9b16648](https://github.com/simplex-chat/simplexmq/pull/547/commits/9b16648cfe3c694302ad706cd9d154fec845f3ec)
Co-authored-by: JRoberts <8711996+jr-simplex@users.noreply.github.com

**Pull request merged**